homeassistant/components/mcp23017/binary_sensor.py

```diff
@@ -85,4 +92,5 @@ def is_on(self):
 
     def update(self):
         """Update the GPIO state."""
-        self._state = self._pin.value
+        with mcp23017Dict:  # Protect device HW access
```

Comment:
We normally don't allow sync locks in our thread pool as that may exhaust the pool. Instead we can make those methods coroutines and use an `asyncio.Lock` that can be awaited without cost.

The device access calls will need to be scheduled on the executor thread pool explicitly with `hass.async_add_executor_job`.

Reply:
Hi Martin,

`asyncio.Lock` is not thread-safe while the issue is related to access from multiple threads to a common HW resource.

If we make this component async, will all instances be called from a single thread ? (I'm new to homeassistant/in discovery mode) If this is the case, then we don't need a lock at all as these series of HW accesses will never be interrupted. 

Cheers,
jpc

Reply:
The async api is always called from within the event loop which is running in the main thread.

Reply:
Hi Martin,

Then I agree that we better convert this component into an async one. The lock won't be required as there is nothing to wait for but a critical section to protect.

@jardiamj, how do you want to proceed here ? (skip this PR and refactor component to async or take it as an intermediate workaround for the issue)

Note that today's code from Adafruit is already blocking the thread:

- https://github.com/adafruit/Adafruit_CircuitPython_BusDevice/blob/master/adafruit_bus_device/i2c_device.py#L167

- Calling a (non thread-safe) lock implementation:
https://github.com/adafruit/Adafruit_CircuitPython_BusDevice/blob/master/adafruit_bus_device/i2c_device.py#L167
 

Cheers
jpc

Reply:
It does sound like refactoring the component to async is the way to go.
My understanding is that since all instances will be called from a single thread, that will solve the threading issue with multiple read/writes happening in parallel.

Reply:
If the library doesn't support asyncio we still need to use our thread pool when interfacing with the library. Preferably the library should support asyncio, then we don't need any locks. Otherwise we may still need an `asyncio.Lock`.

Reply:
The underlying Adafruit libraries aren't based on asyncio but will be nonblocking if they are used within an asyncio context (wrappers aroung I2C read/write). I can have a look at this in a week or two and raise another PR